internal/persistence/sql/migrations/uuid_mapping_test.go

```diff
@@ -0,0 +1,102 @@
+package migrations
+
+import (
+	"context"
+	dbsql "database/sql"
+	"testing"
+	"time"
+
+	"github.com/gobuffalo/pop/v6"
+	"github.com/gofrs/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ory/keto/internal/driver"
+	"github.com/ory/keto/internal/persistence/sql"
+	"github.com/ory/keto/internal/x/dbx"
+)
+
+func TestToUUIDMappingMigrator(t *testing.T) {
+	const debugOnDisk = false
+
+	for _, dsn := range dbx.GetDSNs(t, debugOnDisk) {
+		t.Run("db="+dsn.Name, func(t *testing.T) {
+			ctx := context.Background()
+			r := driver.NewTestRegistry(t, dsn)
+			m := NewToUUIDMappingMigrator(r)
+			p := m.d.Persister().(*sql.Persister)
+			conn := p.Connection(ctx)
+
+			testCases := []struct {
+				name          string
+				rt            *sql.RelationTuple
+				expectMapping bool
+			}{{
+				name: "with string subject",
+				rt: &sql.RelationTuple{
+					ID:         uuid.Must(uuid.NewV4()),
+					Object:     "object",
+					SubjectID:  dbsql.NullString{String: "subject", Valid: true},
+					CommitTime: time.Now(),
+				},
+				expectMapping: true,
+			}, {
+				name: "with null subject",
+				rt: &sql.RelationTuple{
+					ID:                    uuid.Must(uuid.NewV4()),
+					SubjectID:             dbsql.NullString{String: "", Valid: false},
+					SubjectSetNamespaceID: dbsql.NullInt32{Int32: 0, Valid: true},
+					SubjectSetObject:      dbsql.NullString{String: "object", Valid: true},
+					SubjectSetRelation:    dbsql.NullString{String: "subject_set_relation", Valid: true},
+					Object:                "object",
+					CommitTime:            time.Now(),
+				},
+				expectMapping: true,
+			}, {
+				name: "with UUID subject",
+				rt: &sql.RelationTuple{
+					ID:         uuid.Must(uuid.NewV4()),
+					SubjectID:  dbsql.NullString{String: uuid.Must(uuid.NewV4()).String(), Valid: true},
+					CommitTime: time.Now(),
+				},
+				expectMapping: false,
+			}}
+
+			for _, tc := range testCases {
+				t.Run("case="+tc.name, func(t *testing.T) {
+					require.NoError(t, conn.Create(tc.rt))
+					require.NoError(t, m.MigrateUUIDMappings(ctx))
+
+					newRt := &sql.RelationTuple{}
+					require.NoError(t, conn.Find(newRt, tc.rt.ID))
+
+					if tc.expectMapping {
+						// Check subject mapping
+						if tc.rt.SubjectID.Valid {
+							assertHasMapping(t, conn, tc.rt.SubjectID.String, newRt.SubjectID.String)
+						} else {
+							assertHasMapping(t, conn, tc.rt.SubjectSetObject.String, newRt.SubjectSetObject.String)
+							// check that both "OBJECT" strings are mapped to same UUID.
+							assert.Equal(t, newRt.Object, newRt.SubjectSetObject.String)
+						}
+						assertHasMapping(t, conn, tc.rt.Object, newRt.Object)
+					} else {
+						// Nothing should have changed (ignoring commit time)
+						newRt.CommitTime = tc.rt.CommitTime
+						assert.Equal(t, tc.rt, newRt)
+					}
+				})
+			}
```

Comment:
I would like to add a test case here that ensures the pagination works. There is a wrapper for the relation tuple manager where you can overwrite the default pagination: https://github.com/ory/keto/blob/1e15176233eac5db982721e95604cd8e9d47e1a2/internal/relationtuple/definitions.go#L644-L660
You can initialize and use it in the test similar to https://github.com/ory/keto/blob/1e15176233eac5db982721e95604cd8e9d47e1a2/internal/check/engine_test.go#L34-L42
https://github.com/ory/keto/blob/1e15176233eac5db982721e95604cd8e9d47e1a2/internal/check/engine_test.go#L441

Reply:
Thanks, I added a test for pagination. I did not use the `ManagerWrapper`, because I was not retrieving the tuples via the manager anyways, but I used that code as inspiration nonetheless 😉.